crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,623 @@
+    }
+    return ret;
+}
+
+int RSAZ_mod_exp_x2_ifma256(BN_ULONG *out,
+                            const BN_ULONG *base,
+                            const BN_ULONG *exp[2],
+                            const BN_ULONG *m,
+                            const BN_ULONG *rr,
+                            const BN_ULONG k0[2],
+                            int modulus_bitsize)
+{
+    typedef void (*DAMM)(BN_ULONG *res, const BN_ULONG *a,
+                         const BN_ULONG *b, const BN_ULONG *m,
+                         const BN_ULONG k0[2]);
+    typedef void (*DEXTRACT)(BN_ULONG *res, const BN_ULONG *red_table,
+                             int red_table_idx, int tbl_idx);
+
+    int ret = 0;
+    int idx;
+
+    /* Exponent window size */
+    int exp_win_size = 5;
+    int exp_win_mask = (1U << exp_win_size) - 1;
+
+    /*
+    * Number of digits (64-bit words) in redundant representation to handle
+    * modulus bits
+    */
+    int red_digits = 0;
+    int exp_digits = 0;
+
+    BN_ULONG *storage = NULL;
+    BN_ULONG *storage_aligned = NULL;
+    int storage_len_bytes = 0;
+
+    /* Red(undant) result Y and multiplier X */
+    BN_ULONG *red_Y = NULL;     /* [2][red_digits] */
+    BN_ULONG *red_X = NULL;     /* [2][red_digits] */
+    /* Pre-computed table of base powers */
+    BN_ULONG *red_table = NULL; /* [1U << exp_win_size][2][red_digits] */
+    /* Expanded exponent */
+    BN_ULONG *expz = NULL;      /* [2][exp_digits + 1] */
+
+    /* Dual AMM */
+    DAMM damm = NULL;
+    /* Extractor from red_table */
+    DEXTRACT extract = NULL;
+
+/*
+ * Squaring is done using multiplication now. That can be a subject of
+ * optimization in future.
+ */
+# define DAMS(r,a,m,k0) damm((r),(a),(a),(m),(k0))
+
+    switch (modulus_bitsize) {
+    case 1024:
+        red_digits = 20;
+        exp_digits = 16;
+        damm = ossl_rsaz_amm52x20_x2_ifma256;
+        extract = ossl_extract_multiplier_2x20_win5;
+        break;
+    case 1536:
+        // Extended with 2 digits padding to avoid mask ops in high YMM register 
+        red_digits = 30 + 2;
+        exp_digits = 24;
+        damm = ossl_rsaz_amm52x30_x2_ifma256;
+        extract = ossl_extract_multiplier_2x30_win5;
+        break;
+    case 2048:
+        red_digits = 40;
+        exp_digits = 32;
+        damm = ossl_rsaz_amm52x40_x2_ifma256;
+        extract = ossl_extract_multiplier_2x40_win5;
+        break;
+    default:
+        goto err;
+    }
+
+    storage_len_bytes = (2 * red_digits                         /* red_Y     */
+                       + 2 * red_digits                         /* red_X     */
+                       + 2 * red_digits * (1U << exp_win_size)  /* red_table */
+                       + 2 * (exp_digits + 1))                  /* expz      */
+                       * sizeof(BN_ULONG)
+                       + 64;                                    /* alignment */
+
+    storage = (BN_ULONG *)OPENSSL_malloc(storage_len_bytes);
```

Comment:
same question as above, why do we need continuous storage? and can we do stack allocation instead?

Reply:
[This](https://github.com/aws/aws-lc/pull/1273#discussion_r1559735455) applies here as well, I think.